File: ios-core/Sources/ImpalaCore/Workflow.swift
import Foundation

public final class OrderWorkflow {
    public init() {}

    public func assignCourier(order: DeliveryOrder, courierId: String) throws -> DeliveryOrder {
        try ensureTransition(from: order.status, to: .inProgress)
        var next = order
        next.status = .inProgress
        next.courierId = courierId
        next.updatedAt = Date()
        return next
    }

    public func startQrValidation(order: DeliveryOrder) throws -> DeliveryOrder {
        try ensureTransition(from: order.status, to: .waitingQrValidation)
        var next = order
        next.status = .waitingQrValidation
        next.qrToken = QrService.generateToken(order: order)
        next.updatedAt = Date()
        return next
    }

    public func validateQr(order: DeliveryOrder, scannedToken: String) throws -> DeliveryOrder {
        try ensureTransition(from: order.status, to: .waitingPayment)
        guard let token = order.qrToken, !token.isEmpty, token == scannedToken else {
            throw ImpalaError.invalidOrderData("qrToken")
        }
        var next = order
        next.status = .waitingPayment
        next.updatedAt = Date()
        return next
    }

    public func completeAfterPayment(order: DeliveryOrder, payment: PaymentRecord) throws -> DeliveryOrder {
        try ensureTransition(from: order.status, to: .completed)
        var next = order
        next.status = .completed
        next.paymentReference = payment.reference
        next.paymentRecord = payment
        next.tracking.append(
            TrackingEvent(
                eventType: .delivered,
                note: "Livraison validee apres paiement \(payment.reference)",
                position: next.currentLocation
            )
        )
        next.updatedAt = Date()
        return next
    }

    public func updateTracking(order: DeliveryOrder, snapshot: TrackingSnapshot) throws -> DeliveryOrder {
        if order.status == .completed || order.status == .failed {
            throw ImpalaError.invalidStateTransition(from: order.status, to: order.status)
        }
        var next = order
        let point = GeoPoint(latitude: snapshot.latitude, longitude: snapshot.longitude, accuracyMeters: snapshot.accuracyMeters)
        next.currentLocation = point
        next.etaMinutes = snapshot.etaMinutes
        let eventType: TrackingEventType = {
            switch order.status {
            case .pending: return .created
            case .inProgress: return .inTransit
            case .waitingQrValidation, .waitingPayment: return .arrivedAtDestination
            case .completed: return .delivered
            case .failed: return .inTransit
            }
        }()
        next.tracking.append(
            TrackingEvent(eventType: eventType, note: "Tracking update", position: point)
        )
        next.updatedAt = Date()
        return next
    }

    private func ensureTransition(from: OrderStatus, to: OrderStatus) throws {
        let allowed: [OrderStatus]
        switch from {
        case .pending:
            allowed = [.inProgress]
        case .inProgress:
            allowed = [.waitingQrValidation]
        case .waitingQrValidation:
            allowed = [.waitingPayment]
        case .waitingPayment:
            allowed = [.completed, .failed]
        case .completed, .failed:
            allowed = []
        }

        if !allowed.contains(to) {
            throw ImpalaError.invalidStateTransition(from: from, to: to)
        }
    }
}

public enum QrService {
    public static func generateToken(order: DeliveryOrder) -> String {
        let raw = "\(order.id.uuidString)|\(order.orderNumber)|\(order.recipientPhoneNumber)"
        return simpleHash(raw).prefix(32).description
    }

    private static func simpleHash(_ text: String) -> String {
        let bytes = [UInt8](text.utf8)
        var hash = [UInt8](repeating: 0, count: 32)
        for (index, byte) in bytes.enumerated() {
            let slot = index % 32
            hash[slot] = hash[slot] &+ byte &+ UInt8(slot)
        }
        return hash.map { String(format: "%02x", $0) }.joined()
    }
}

File: android-app/shared-core/src/main/kotlin/com/impala/core/Workflow.kt
package com.impala.core

import java.nio.charset.StandardCharsets
import java.security.MessageDigest
import java.time.Instant

sealed class OfflineAction {
    data class CreateOrder(val orderId: String) : OfflineAction()
    data class UpdateStatus(
        val orderId: String,
        val status: OrderStatus,
        val courierId: String? = null,
        val qrToken: String? = null,
        val paymentReference: String? = null
    ) : OfflineAction()
    data class TrackingUpdate(
        val orderId: String,
        val location: TrackingLocation
    ) : OfflineAction()
}

enum class OperationResult {
    SUCCESS,
    RETRY
}

class OfflineSyncQueue {
    private val actions = mutableListOf<OfflineAction>()

    fun enqueue(action: OfflineAction) {
        actions += action
    }

    fun drain(): List<OfflineAction> {
        val snapshot = actions.toList()
        actions.clear()
        return snapshot
    }

    fun size(): Int = actions.size
    fun isEmpty(): Boolean = actions.isEmpty()

    fun flush(executor: (OfflineAction) -> OperationResult) {
        val failed = mutableListOf<OfflineAction>()
        actions.forEach { action ->
            val result = executor(action)
            if (result == OperationResult.RETRY) {
                failed += action
            }
        }
        actions.clear()
        actions += failed
    }
}

class OrderWorkflow {
    fun assignCourier(order: DeliveryOrder, courierId: String): DeliveryOrder {
        ensureTransition(order.status, OrderStatus.IN_PROGRESS)
        return order.copy(
            status = OrderStatus.IN_PROGRESS,
            courierId = courierId,
            updatedAt = Instant.now()
        )
    }

    fun startQrValidation(order: DeliveryOrder): DeliveryOrder {
        ensureTransition(order.status, OrderStatus.WAITING_QR_VALIDATION)
        return order.copy(
            status = OrderStatus.WAITING_QR_VALIDATION,
            qrToken = QrService.generateToken(order),
            updatedAt = Instant.now()
        )
    }

    fun validateQr(order: DeliveryOrder, scannedToken: String): DeliveryOrder {
        ensureTransition(order.status, OrderStatus.WAITING_PAYMENT)
        if (order.qrToken.isNullOrBlank() || scannedToken != order.qrToken) {
            throw DomainError.InvalidOrderData("qrToken")
        }
        return order.copy(
            status = OrderStatus.WAITING_PAYMENT,
            updatedAt = Instant.now()
        )
    }

    fun completeAfterPayment(order: DeliveryOrder, payment: PaymentRecord): DeliveryOrder {
        ensureTransition(order.status, OrderStatus.COMPLETED)
        return order.copy(
            status = OrderStatus.COMPLETED,
            paymentReference = payment.reference,
            paymentRecord = payment,
            tracking = order.tracking + TrackingEvent(
                eventType = TrackingEventType.DELIVERED,
                note = "Livraison validee apres paiement ${payment.reference}",
                position = order.currentLocation
            ),
            updatedAt = Instant.now()
        )
    }

    fun updateTracking(order: DeliveryOrder, snapshot: TrackingSnapshot): DeliveryOrder {
        if (order.status == OrderStatus.COMPLETED || order.status == OrderStatus.FAILED) {
            throw DomainError.InvalidStateTransition(order.status, order.status)
        }
        val eventType = when (order.status) {
            OrderStatus.PENDING -> TrackingEventType.CREATED
            OrderStatus.IN_PROGRESS -> TrackingEventType.IN_TRANSIT
            OrderStatus.WAITING_QR_VALIDATION -> TrackingEventType.ARRIVED_AT_DESTINATION
            OrderStatus.WAITING_PAYMENT -> TrackingEventType.ARRIVED_AT_DESTINATION
            OrderStatus.COMPLETED -> TrackingEventType.DELIVERED
            OrderStatus.FAILED -> TrackingEventType.IN_TRANSIT
        }
        val point = GeoPoint(snapshot.latitude, snapshot.longitude)
        return order.copy(
            currentLocation = point,
            etaMinutes = snapshot.etaMinutes,
            tracking = order.tracking + TrackingEvent(
                eventType = eventType,
                note = "Tracking update",
                position = point
            ),
            updatedAt = Instant.now()
        )
    }

    private fun ensureTransition(from: OrderStatus, to: OrderStatus) {
        val allowed = when (from) {
            OrderStatus.PENDING -> setOf(OrderStatus.IN_PROGRESS)
            OrderStatus.IN_PROGRESS -> setOf(OrderStatus.WAITING_QR_VALIDATION)
            OrderStatus.WAITING_QR_VALIDATION -> setOf(OrderStatus.WAITING_PAYMENT)
            OrderStatus.WAITING_PAYMENT -> setOf(OrderStatus.COMPLETED, OrderStatus.FAILED)
            OrderStatus.COMPLETED, OrderStatus.FAILED -> emptySet()
        }
        if (to !in allowed) {
            throw DomainError.InvalidStateTransition(from, to)
        }
    }
}

data class TrackingSnapshot(
    val latitude: Double,
    val longitude: Double,
    val speedKmh: Double? = null,
    val etaMinutes: Int? = null
)

object QrService {
    fun generateToken(order: DeliveryOrder): String {
        val raw = "${order.id}|${order.orderNumber}|${order.recipientPhoneNumber}"
        val digest = MessageDigest.getInstance("SHA-256")
            .digest(raw.toByteArray(StandardCharsets.UTF_8))
        return digest.joinToString("") { "%02x".format(it) }.take(32)
    }
}
